lib/active_merchant/billing/gateways/flex_charge.rb

```diff
@@ -0,0 +1,327 @@
+          firstName: first_name,
+          lastName: last_name,
+          country: address[:country],
+          phone: address[:phone],
+          countryCode: address[:country],
+          addressLine1: address[:address1],
+          state: address[:state],
+          city: address[:city],
+          zipCode: address[:zip]
+        }
+      end
+
+      def add_invoice(post, money, options)
+        post[:transaction] = {
+          id: options[:order_id],
+          dynamicDescriptor: options[:dynamic_descriptor],
+          timestampUtc: options[:timestamp_utc] || Time.now,
+          timezoneUtcOffset: options[:timestamp_utc_offset],
+          amount: money,
+          currency: (options[:currency] || currency(money)),
+          responseCode: options[:response_code],
+          responseCodeSource: options[:response_code_source] || 'NA',
+          avsResultCode: options[:avs_result_code],
+          cvvResultCode: options[:cvv_result_code],
+          cavvResultCode: options[:cavv_result_code] || 'NA',
+          cardNotPresent: options[:card_not_present]
+        }
+      end
+
+      def add_transaction_type(post, action)
+        return unless %i(authorize capture void).include?(action)
+
+        post[:transaction] ||= {}
+        post[:transaction][:transactionType] =
+          case action
+          when :authorize
+            'Auth'
+          when :capture
+            'Capture'
+          when :void
+            'Void'
+          end
+      end
+
+      def add_payment(post, credit_card, address, options)
+        post[:paymentMethod] = {
+          holderName: credit_card.name,
+          cardType: options.dig(:payment_token, :card_type)&.upcase || 'CREDIT',
+          cardBrand: credit_card.brand&.upcase,
+          cardCountry: options.dig(:payment_token, :country) || address[:country],
+          expirationMonth: options.dig(:payment_token, :month) || credit_card.month,
+          expirationYear: options.dig(:payment_token, :year) || credit_card.year,
+          cardBinNumber: options.dig(:payment_token, :first_six_digits) || credit_card.number[0..5],
+          cardLast4Digits: options.dig(:payment_token, :last_four_digits) || credit_card.number[-4..-1],
+          cardNumber: options.dig(:payment_token, :token) || credit_card.number
+        }
+      end
+
+      def add_credit_card(post, credit_card, options)
+        first_name, last_name = split_names(credit_card.name)
+        post[:payment_method] = {
+          sense_key: options[:order_id],
+          credit_card: {
+            first_name: first_name,
+            last_name: last_name,
+            number: credit_card.number,
+            verification_value: credit_card.verification_value,
+            month: credit_card.month,
+            year: credit_card.year
+          }
+        }
+      end
+
+      def add_post_data(post)
+        post[:mid] = @options[:mid]
+        post[:environment] = @options[:tokenization_key] || @options[:access_token]
```

Comment:
❓ Question:

I can't find the `environment` parameter and I can't infer how the tokenization_key or access_token are related to it.

Reply:
Fixed!